### PR TITLE
Custom stopwords

### PR DESCRIPTION
**Pull request by cmacdonald**
Closes: #361 

This addresses the request in #361 for the ability to change the stopword list at indexing time. 

The indexing API then looks like:
```python
indexer = pt.IterDictIndexer("./index", stopwords=['a', 'an', 'the'])
```

We need to look more at how good the documentation is.

NB: I dont expect the automated tests to pass until various things are merged, committed etc. 
Requires:
 - merging of https://github.com/terrier-org/terrier-core/pull/218
 - release of terrier-core
 - release of new terrier-python-helper

### CODE REVIEW

**Comment by cmacdonald** [comment]
if a filename or URL is provided, should TerrierStopwords behave in a similar way, i.e. by ingesting the file and copying it into the data.properties file to act as a stopwords list?

File: tests/test_iterdictindex.py [hunk 1]
```diff
@@ -190,6 +190,31 @@ def test_check_nostemmer_utf_toks(self):
         # restore setting after test
         pt.ApplicationSetup.setProperty("termpipelines", "Stopwords,PorterStemmer")
 
+    def test_check_customstops(self):
+        it = [
+            {'docno': '1', 'url': 'url1', 'text': 'He ran out of money, so he had to stop playing', 'title': 'Woes of playing poker'},
+            {'docno': '2', 'url': 'url2', 'text': 'The waves were crashing on the shore; it was a', 'title': 'Lovely sight'},
+            {'docno': '3', 'url': 'url3', 'text': 'The body may perhaps compensates før the loss', 'title': 'Best of Viktor Prowoll'},
+        ]
+        indexer = pt.IterDictIndexer(self.test_dir, stemmer=pt.TerrierStemmer.none, stopwords=['money', 'crashing'], overwrite=True)
```
Comment (seanmacavaney) [inline tests/test_iterdictindex.py:199]: Can we add a test with UTF characters in the stopwords?

Is there anything that's invalid that should raise an error before causing havoc? Probably `,` and `\n`? Maybe others?

**Review by seanmacavaney — CHANGES_REQUESTED** [review]
A few corner cases to consider, commented in code. I'm not much a fan of setting the (global) `TerrierStopwords.custom.stopwords`

The docstring and typing should be updated in TerrierIndexer to indicate that the stopwords parameter can take `List[str]` too.

File: pyterrier/index.py [hunk 2]
```diff
@@ -325,14 +325,47 @@ class TerrierStopwords(Enum):
 
     none = 'none' #: No Stemming
     terrier = 'terrier' #: Apply Terrier's standard stopword list
+    custom = 'custom' #: Apply PyTerrierCustomStopwordList.Indexing for indexing, and PyTerrierCustomStopwordList.Retrieval for retrieval
 
     @staticmethod
     def _to_obj(this):
+        if isinstance(this, list):
+            rtr = TerrierStopwords('custom')
+            rtr.stopwords = list(this)
```
[the remark this answers is not in the document]
Comment (cmacdonald) [reply pyterrier/index.py:334]: This appears to be a misunderstanding on my part. Suggestions on how to improve?

File: terrier-python-helper/pom.xml [hunk 3]
```diff
@@ -5,7 +5,7 @@
 
   <groupId>org.terrier</groupId>
   <artifactId>terrier-python-helper</artifactId>
-  <version>0.0.7</version>
+  <version>0.0.8-SNAPSHOT</version>
```
[the remark this answers is not in the document]
Comment (cmacdonald) [reply terrier-python-helper/pom.xml:8]: Agreed; the Java code to be /released/ before this PR is merged.

File: tests/test_iterdictindex.py (hunk 1 above)
Comment (cmacdonald) [reply tests/test_iterdictindex.py:199]: i added a fix for comma. whitespace is rarely a token anywhere.

File: pyterrier/index.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (cmacdonald) [reply pyterrier/index.py]: now fixed

**Comment by cmacdonald** [comment]
I ran tests on another machine as follows:

```bash
git clone git@github.com:terrier-org/terrier-core.git
cd terrier-core
git checkout --track origin/index_configurable_termpipelines
mvn clean; mvn -DskipTests install
cd ..
git clone git@github.com:terrier-org/pyterrier.git
cd pyterrier
git checkout --track origin/custom_stopwords
cd terrier-python-helper && mvn clean &&  mvn -DskipTests install
TERRIER_VERSION=5.8-SNAPSHOT TERRIER_HELPER_VERSION=0.0.8-SNAPSHOT pytest 
```

and this ran fine.

**Review by tonellotto — COMMENTED** [review]
In my opinion, we need an additional test: given an index with a custom stopwords list in .properties, check if queries are correctly processed as expected.

**Comment by cmacdonald** [comment]
Thanks, will give a look.

**Comment by cmacdonald** [comment]
Update - the necessary test is there :-)

**Pull request merged**